### ISSUE TITLE
Identify the correct controller behavior for pausing / resuming in automated canary release pipelines

### ISSUE THREAD

**Issue by sriumcp** [iter8-tools/iter8-controller#222]
Description of the problem:
```
There was only thing I noticed and not sure if its a bug or just the message is confusing. The first time I deployed I did not realise the experiment was going to run. I then triggered v2 deployment which created the next experiment. Because the first experiment was running the second one went into a paused state. I expected the second experiment to start once the first one was complete. It had not started after 15mins of the first experiment completing. I could not see a way of unpausing the experiment. I had to remove my v2 chart and re-deploy to start again. Should the experiment have been marked as stopped or canceled or would it have started up if I left it longer?
```

**Comment by kalantar** [comment]
We should recognize that this is a special case of something that often happens in a pipeline: we start a new experiment before a previous one is complete. This happens, for example, when we merge code resulting in a canary experiment immediately after another merge. The result is a two canary experiments. Since we hard code the values of baseline and candidate, both experiments will indicate the _same_ baseline. However, since the first is not completed, the second will be paused. When the first finishes, the second may begin _but_ note that a successful first experiment will have changed the desired baseline -- our experiment definition is now invalid and if we use it we will do the wrong thing.

In the case Daniel identifies, this isn't an issue since baseline and candidate of the first experiment are the same. Restarting the paused experiment will work in this case, but will fail in the general case. We should consider (and expect) the general case.  @fabolive it might be an idea to describe the general problem to Daniel and get feedback on whether or not he thinks this will be an issue. I do since they are talking about automation.  I believe @mnuttall and Duane also identified this issue.

**Comment by sriumcp** [comment]
@kalantar I believe the issue is important enough for us to solve. Clearly, users need to be able to use a *single, fixed* automation pipeline (e.g., written in Tekton or TravisCI) and be able to do multiple experiments over time automatically.

We are solving the automation problem within the context of a single experiment now, IMO, it is critical that we address this lifecycle issue, not just for this particular user but for potential future users.

**Comment by sriumcp** [comment]
I also want to split this problem into two situations.

**Initialization:** I realize one of the desirable side-effects of an iter8 experiment is the creation of a virtual service which can be used for bootstrapping the first release. But I don't think this is the right way to think about / use an experiment. Perhaps, we can bundle an **(helm templatized?) virtual service** for bootstrapping that people can use in the first stage of their pipeline (when there is a need to instantiate the VS for a first time)?

I think it is important to keep the iter8 experiment's purpose self-contained and simple -- which is to conduct an online field experiment using multiple versions of a service (including the degenerate case supported in v0.2 of using the same baseline and canary). If the user didn't intend to actually perform an experiment, but only create a VS -- I fully understand, but we need to support this in a different way, not through the experiment creation mechanism.

**Comment by sriumcp** [comment]
@kimikowang I remember you tagging deployments in the past with iter8's labels, and not sure if this mechanism is still being used by the controller today. 

Is the use of appropriate deployment / service labels (which can be used by the controller to determine the appropriate baseline / candidates in an experiment) the way to approach this problem?

**Comment by kimikowang** [comment]
> @kimikowang I remember you tagging deployments in the past with iter8's labels, and I am not sure if this mechanism is still being used by the controller today.
> 
> Is the use of appropriate deployment / service labels (which can be used by the controller to determine the appropriate baseline / candidates in an experiment) the way to approach the **steady-state** problem?

No, labels are only applied to routing rules now.

**Comment by sriumcp** [comment]
Ok. What is the way forward for us in terms of being able to **not** hardcode baseline identifiers in experiment CR? If iter8 were to manipulate deployment / service labels at the end of an experiment, will it enable us to solve this issue?

Of course, we will need to make (and clearly state) the assumption that the labels created by iter8 are not being overwritten / deleted out-of-band.

**Comment by kalantar** [comment]
I'm not disagreeing that keeping things separate would be better. However, the difficulty with this split of initialization and of steady state is that it doesn't help address the pipeline automation problem. It makes the assumption that someone will do something different in different scenarios. This is precisely what causes difficulties. Of course to keep the concepts of iter8 clean we would want to avoid a degenerate case like common baseline/candidate. But to keep life simple for the lifecycle, it is helpful, IMO, for all functions (not just iter8) to support degenerate cases with reasonable behavior.  And yes, it would be helpful if automation tools supported better conditional logic too.

**Comment by sriumcp** [comment]
@kalantar Is it possible to write the automation pipeline as follows...

Check VS is already present. 
1) No? You're in the initialization phase -- create VS and move on.
2) Yes? You're in the steady state phase -- create experiment and move on.

So, we still have a single pipeline, but internally, it does different things conditionally, allowing **iter8** to maintain proper abstractions.

**Comment by sriumcp** [comment]
If we agree the above is the correct abstraction in terms of iter8 and its usage, then we absolutely should be recommending to our users that they write the automation pipeline in the above way (and not a way that forces iter8 to do weird things, like create an experiment only for the sake of its side-effect of creating a VS and with no intention of performing the experiment!)

**Comment by sriumcp** [comment]
I have another thought / proposal in this context, which involves using an experiment life-cycle CRD. We had briefly discussed this idea long ago... Here it is.

1. Keep the experiment CR as it stands right now -- specifically, baseline and candidate(s) are hard coded using identifiers. The experiment CR exclusively focuses on what needs to happen during a **single** iter8 experiment.

2. Define a new CRD called experiment life-cycle CRD. The life-cycle describes how new experiments are created and parameterized. For e.g., which experiments can proceed in parallel, when should a new experiment spec be created, should it be queued, when can an experiment be dequeued and run, how to parameterize the next experiment that is about to be launched -- specifically, how to fill in baseline and candidate values within this experiment, and so on. The policies that dictate these decisions, **spanning across multiple experiments** belong to the life-cycle CRD.

I believe life-cycle CRD is the right abstraction for automated pipelines.

I also believe the experiment CR is the right abstraction for many users who want to automate individual experiments, but not necessarily go down the pipeline route.. in other words, between on experiment and the next, they will take control. Obviously, this is an important building block that will be used by the life-cycle CRD as well.

The main reason I am proposing this is to keep the experiment CRD abstraction as simple and as self-contained as possible, and separate the concerns of a single experiment and multiple successive experiments as cleanly as possible.

**Comment by kimikowang** [comment]
> I have another thought / proposal in this context, which involves using an experiment life-cycle CRD. We had briefly discussed this idea long ago... Here it is.
> 
> 1. Keep the experiment CR as it stands right now -- specifically, baseline and candidate(s) are hard coded using identifiers. The experiment CR exclusively focuses on what needs to happen during a **single** iter8 experiment.
> 2. Define a new CRD called experiment life-cycle CRD. The life-cycle describes how new experiments are created and parameterized. For e.g., which experiments can proceed in parallel, when should a new experiment spec be created, should it be queued, when can an experiment be dequeued and run, how to parameterize the next experiment that is about to be launched -- specifically, how to fill in baseline and candidate values within this experiment, and so on. The policies that dictate these decisions, **spanning across multiple experiments** belong to the life-cycle CRD.
> 
> I believe life-cycle CRD is the right abstraction for automated pipelines.
> 
> I also believe the experiment CR is the right abstraction for many users who want to automate individual experiments, but not necessarily go down the pipeline route.. in other words, between on experiment and the next, they will take control. Obviously, this is an important building block that will be used by the life-cycle CRD as well.
> 
> The main reason I am proposing this is to keep the experiment CRD abstraction as simple and as self-contained as possible, and separate the concerns of a single experiment and multiple successive experiments as cleanly as possible.

I agree that this is a good idea, but this would require another controller for the new CRD. This could introduce complexity to the controller manager(interaction between two controllers). Maybe we can put this into plan, just not the top priority for now.

**Comment by sriumcp** [comment]
@kimikowang Agreed. This can be dealt with in v1.1.0 or later. I think there is at least one more CRD we will need to deal with, namely, metric CRD. Also v1.1.0 or later of course.
